urlc5fdb179cc_lagodiuk_evo_neural_network_agents_tgz-pJ8-com_lagodiuk_agent_evolution_MainJ8/src/com/lagodiuk/nn/serializing/xml/MapAdapter.java

```diff
@@ -0,0 +1,56 @@
+/*******************************************************************************
+ * Copyright 2012 Yuriy Lagodiuk
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+package com.lagodiuk.nn.serializing.xml;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import javax.xml.bind.annotation.adapters.XmlAdapter;
+
+public class MapAdapter extends XmlAdapter<MapElement[], Map<Integer, Map<Integer, Double>>> {
+
+	@Override
+	public MapElement[] marshal(Map<Integer, Map<Integer, Double>> map) throws Exception {
+		List<MapElement> list = new LinkedList<MapElement>();
+
+		for (Integer from : map.keySet()) {
+			Map<Integer, Double> toMap = map.get(from);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.